Storage chain: Runtime module

arkpar opened the pull request:
This implements runtime support for storage chains (#7962).

Each transaction is converted to a trie containing 256 byte chunks. Runtime maintains storage roots for these tries and checks proofs for randomly selected chunk in a past block. Randomness is based on the parent block hash currently.

A couple of issues still left unresolved, to be addressed in a later PR.

* Random chunk selection is currently suboptimal. It requires iterating over all transactions in the block to find a chunk with selected random index. This can be replaced with a binary search at  least. Another possibility is to replace uniform random chunk selection with selecting a random transaction first, and then a random chunk within that transaction.
* Storage period is currently configurable within the runtime, but it must match the hardcoded value int Inherent provider and block pruning for block authoring to work as expected. It is not clear yet weather we need configurable storage period yet.
* Fees are simply burned.

polkadot companion: paritytech/polkadot#2890

bkchr commented on primitives/storage-proof/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Debug)]
+pub struct StorageProof {
```
Can we please rename this crate and this struct? We already got the `StorageProof` struct.

arkpar replied:
Done

bkchr commented on primitives/io/src/lib.rs:
```diff
@@ -427,6 +427,24 @@ pub trait Trie {
 	}
+
+	/// Verify trie proof
+	fn blake2_256_verity_proof(root: H256, proof: &[Vec<u8>], key: &[u8], value: &[u8]) -> bool {
```
`verify`

Here and below.

cheme commented on primitives/transaction-storage-proof/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+
+/// An utility function to enocde transaction index as trie key.
+pub fn encode_index(input: u32) -> Vec<u8> {
+	codec::Encode::encode(&codec::Compact(input))
```
From memory, this encode_index function is a bit awkward (it is the same with the sp_trie definition, but cannot really be change), because Compact encoding of u32 is not always ordered in the trie (protobuf varints are ok for this I think).
I did not read the whole PR, but regarding this proof, it was thinking a mmr or a binary tree could be more compact (we got mmr somewhere in the codebase in frame, and the primitive module used there could really be substrate primitive (just the bagging implementation seems a bit uncertain to me (I would have done the hashes the other way))).

arkpar replied:
That's a neat idea, but I'll save it for a future PR to avoid further complicating this one. Single transaction is currently limited at 32768 trie entries, which limits the proof size to about 1.8 kbytes. I've roughly estimated the MMR proof size to be about 1.4 kbytes for the same number of entries in the tree.  Given that the block size is expected to be in the megabytes, saving a few hundred bytes per block is not that important. Checking MMR proof could be faster however.

kianenigma commented on frame/benchmarking/src/lib.rs:
```diff
@@ -791,6 +794,20 @@ macro_rules! impl_benchmark {
 								"Read/Write Count {:?}", read_write_count
 							);
 
+							let time = $crate::benchmarking::current_time();
```
ahh this is so good 👍

arkpar replied:
It is called outside of the timed benchmark section

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+pub const MAX_DATA_SIZE: u32 = 8 * 1024 * 1024;
+
+#[derive(Encode, Decode, sp_runtime::RuntimeDebug, PartialEq, Eq)]
+pub struct TransactionInfo {
```
doc for public item would be nice.

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// This will be cleared in in `on_finalize`
+			<Counter<T>>::put(0);
+			0
```
here you should return the combined weight of `on_initialize` and `on_finalize`

arkpar replied:
Done

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// This will be cleared in in `on_finalize`
+			<Counter<T>>::put(0);
```
Also I guess since you kill this `on_finalize`, and since default for any num type is already 0, you can simply ignore this write.

arkpar replied:
Done

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+			sp_io::transaction_index::index(extrinsic_index, data.len() as u32, content_hash);
+
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			let mut counter = <Counter<T>>::get().unwrap_or(0);
+			<TransactionRoots<T>>::insert(block_num, counter, TransactionInfo {
+				chunk_root: root,
+				size: data.len() as u32,
+				content_hash: content_hash.into(),
+			});
+			Self::deposit_event(Event::Stored(counter));
+			counter += 1;
+			<Counter<T>>::put(counter);
```
```suggestion
			<TransactionRoots<T>>::insert(block_num, counter, TransactionInfo {
				chunk_root: root,
				size: data.len() as u32,
				content_hash: content_hash.into(),
			});
			let counter = <Counter<T>>::mutate(|c| { *c += 1; c }); 
			Self::deposit_event(Event::Stored(counter));
```

optional

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			<TransactionRoots<T>>::insert(block_num, counter, info);
+			Self::deposit_event(Event::Renewed(counter));
+			counter += 1;
```
ditto

kianenigma commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,354 @@
+	pub(super) type ByteFee<T: Config> = StorageValue<_, BalanceOf<T>>;
+
+	#[pallet::storage]
+	pub(super) type Counter<T: Config> = StorageValue<_, u32>;
```
```suggestion
	pub(super) type Counter<T: Config> = StorageValue<_, u32, ValueQuery>;
```

I think with this you'll get the default for `u32` in case of nothing being there.

kianenigma commented on frame/transaction-storage/src/mock.rs:
```diff
@@ -0,0 +1,148 @@
+	type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>;
+	pub type Block = frame_system::mocking::MockBlock<Test>;
+
+
```
```suggestion
```

kianenigma commented on frame/transaction-storage/src/mock.rs:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	parameter_types! {
+		pub IgnoredIssuance: u64 = Balances::total_balance(&0); // Account zero is ignored.
```
why?

arkpar replied:
Copy-paste leftovers. Removed.

kianenigma commented on frame/transaction-storage/src/mock.rs:
```diff
@@ -0,0 +1,148 @@
+			pallet_balances: pallet_balances::GenesisConfig::<Test> {
+				balances: vec![(1, 1000000000), (2, 100), (3, 100), (4, 100)]
+			},
+			pallet_transaction_storage: pallet_transaction_storage::GenesisConfig::default(),
```
```suggestion
			pallet_transaction_storage: Default::default(),
```

kianenigma commented on frame/transaction-storage/src/mock.rs:
```diff
@@ -0,0 +1,148 @@
+	}
+}
+
+pub fn setup<T: Config>() -> Result<(), &'static str> {
```
This should just be part of your `new_test_ext`?

kianenigma replied:
also unclear why be generic over `T` here.

kianenigma replied:
Okay -- I see that you are using this in benchmarks as well. Makes sense. 

But note that I still think it is a bad pattern. You are now overriding the fee to `2` in all benchmarks, regardless of it being test or the real benchmarks. In fact, in real benchmarks (i.e. run against `substrate-node` or `polkadot`) `2` as balance type is pretty much insubstantial and below ED. 

I think the correct thing is that you should not manually setup these things in your benchmarks. Then, 

1. For testing benchmarks, put them in `new_test_ext()`
2. For real benchmarks, assume that they chain that is running the benchmarks already has sensible value. You can assert to make sure the values are sane though (`assert!(fee > ED)` etc).

arkpar replied:
Done

cheme commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			};
+
+			ensure!(
+				sp_io::trie::blake2_256_verify_proof(
```
So the purpose is to ensure users keep a copy of the storage proof for at least the storage period (somehow similar to having it in state but only by random sampling)?
If it is the case, then we need to write something in state: eg hash(previousblock ++ proof.chunk) so skipping will be a bit more difficult (need to trust a hash from state that we don't know).
(here just need to recompile the wasm to skip the check)

arkpar replied:
Not sure what you mean here. Even if you recompile your node code, the rest of the network will still reject the block that does not contain the proof. The `check_proof` inherent is mandatory.

cheme replied:
I think I misunderstood how this work: the 'proof' parameter will be included in the block? (I was thinking every client would fetch their own storage for it).
So this will motivate only block producer to keep this content.

cheme replied:
(the scenario I was thinking of would have been implemented with a host function in fact (not an inherent)): calls to host function to access the data then hash it with prev block hash, so all network need the chunk sample (or fetch it somehow, which makes it a middleground between storage and speed and should probably need to query more than one chunk) and then check block from it.
What prevent user to skip this check is the inhability to build the next state.

Clearly, at this point I just do not really understand the purpose of this check, my hypothesis is checking that block producer keep access to this content.

cheme replied:
The chunk could be added to the proof the same way trie nodes are added, but that is certainly more impacting.

cheme commented on primitives/transaction-storage-proof/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+	let mut buf = [0u8; 8];
+	buf.copy_from_slice(&random_hash[0..8]);
+	let random_u64 = u64::from_be_bytes(buf);
+	(random_u64 % total_chunks as u64) as u32
```
Should we configure a number of checked chunk (not only one) depending on the total _chuncks?

arkpar replied:
1 should be good enough to discourage validators from messing around with data and potentially losing bock rewards. We can increase it later if required.

cheme commented on primitives/io/src/lib.rs:
```diff
@@ -427,6 +427,24 @@ pub trait Trie {
 	}
+
+	/// Verify trie proof
+	fn blake2_256_verify_proof(root: H256, proof: &[Vec<u8>], key: &[u8], value: &[u8]) -> bool {
```
Maybe:
```suggestion
	fn blake2_256_verify_proof(root: H256, proof: &[Vec<u8>], key: &[&[u8]], value: &[&[u8]]) -> bool {
```

arkpar replied:
That's not supported by the `runtime_interface` macro unfortunately

cheme commented on primitives/transaction-storage-proof/src/lib.rs:
```diff
@@ -0,0 +1,243 @@
+							}
+						};
+
+						let number = chain_head.number().saturating_sub(DEFAULT_STORAGE_PERIOD.into());
```
Is it fine to use directly 'DEFAULT_STORAGE_PERIOD' or should we fetch it from the runtime module (at this time nothing change it but it could?)

arkpar replied:
As I've mentioned in the PR description, that's undecided for now. We'll probably remove storage period as config option for now.

cheme reviewed (commented):
I had a look at the rest of the PR, some point may be a bit wrong since I am not a runtime specialist.

But the runtime aspect were really interesting, I guess it makes the final design of 'storage chain' way more explicit as seems to adds constraint for nodes to keep the storage blob (I feel like we should store hash(parent_hash ++ chunk) additionally).
So the point IIUC is to assert the user did keep the storage.
Is there design/plan for a protocol to ensure the user also serves it?

This PR also makes me realize that 'TransactionStorage' can be a long name ('pallet_transaction_storage'), maybe finding a short name for it could be interesting, but I got no good idea.

arkpar commented on client/db/src/lib.rs:
```diff
@@ -594,6 +594,37 @@ impl<Block: BlockT> sc_client_api::blockchain::Backend<Block> for BlockchainDb<B
+
+	fn block_indexed_body(&self, id: BlockId<Block>) -> ClientResult<Option<Vec<Vec<u8>>>> {
+		match self.transaction_storage {
+			TransactionStorageMode::BlockBody => Ok(None),
```
There's no default notion for indexed data. `BlockBody` means that no indexed data is stored, so `None` looks appropriate. Also it would be incorrect to return `Some` for blocks that are truly unknown.

arkpar commented on frame/transaction-storage/src/benchmarking.rs:
```diff
@@ -0,0 +1,170 @@
+		let l in 1 .. MAX_DATA_SIZE;
+		let caller: T::AccountId = whitelisted_caller();
+		T::Currency::make_free_balance_be(&caller, BalanceOf::<T>::max_value());
+	}: _(RawOrigin::Signed(caller.clone()), vec![0u8; l as usize])
```
What do you mean? This benchmarks the whole of `store` function, including all externalities that are called from it.

cheme replied:
The question is, does it account to the fact that the extrinsic is not stored as a single value but two value, since this storage cost is more a client thing, I do not know if benchmarking will see that?

cheme replied:
Replied in an other comment by:
> We don't include costs that are outside of the runtime execution to the weights I believe.
I do not really know what to think of it (I am really not the bench specialist).

shawntabrizi replied:
What other operations are happening here outside the runtime execution?

Weights are for making sure that a full block can be imported in time.

Will any of the operations affect block import for other nodes syncing?

arkpar replied:
Well, weights don't account for putting the block body in the database, updating change tries, sending notifications, pruning transaction pool, etc. The extra step to put indexed body as two DB values instead of one won't add much to it and is constant time.

arkpar commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		/// - n*log(n) of data size, as all data is pushed to an in-memory trie.
+		/// Additionally contains a DB write.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::store(data.len() as u32))]
```
The comment is a bit outdated. However we don't include costs that are outside of the runtime execution to the weights I believe.

shawntabrizi replied:
if it would increase the block import time, we should

arkpar commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
```
I was asked to remove it in https://github.com/paritytech/substrate/pull/8624#discussion_r618973135

cheme replied:
Damn (sorry), I am thinking it is better with, but not very important.

arkpar commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		/// # <weight>
+		/// - Constant with a single DB read.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::renew())]
```
The comment is a bit outdated. `T::WeightInfo::renew()` actually does that

arkpar commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			let chunks = data.chunks(CHUNK_SIZE).map(|c| c.to_vec()).collect();
+			let root = sp_io::trie::blake2_256_ordered_root(chunks);
+
+			let content_hash = sp_io::hashing::blake2_256(&data);
```
This could be a nice optimization, but I'll save it for later, once the MVP is working.

arkpar replied:
See https://github.com/paritytech/substrate/pull/8624/files#r642780466

arkpar commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn apply_fee(origin: OriginFor<T>, size: u32) -> DispatchResult {
```
It does return `DispatchResult` which makes it obvious enough that it may fail to apply fee for whatever reason. No signature or not enough balance namely.

arkpar commented:
@kianenigma @bkchr @cheme I think all issues have been addressed here. Waiting on your approval (and don't forget the companion)

cheme reviewed (approved):
Wanted to do it two days ago, but had been stuck on something else and was delaying for no reason.

shawntabrizi commented on frame/transaction-storage/README.md:
```diff
@@ -0,0 +1,2 @@
+
```
whats up with this suggestion?

arkpar replied:
Already merged

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+
+/// Maximum bytes that can be storead in one transaction.
+// Increasing it further also requires raising the allocator limit.
+pub const MAX_DATA_SIZE: u32 = 8 * 1024 * 1024;
```
Should this be something that we allow the runtime developer to configure?

For example, someone wants a chain with more or less than this?

arkpar replied:
As per commend above, they would also need to increase the hardcoded allocator limit.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+#[derive(Encode, Decode, sp_runtime::RuntimeDebug, PartialEq, Eq)]
+pub struct TransactionInfo {
+	/// Chunk trie root.
+	chunk_root: <BlakeTwo256 as Hash>::Output,
+	/// Plain hash of indexed data.
+	content_hash: <BlakeTwo256 as Hash>::Output,
```
Should we not use the `T::Hash` type from `frame_system`?

Is there any reason we def want to always use blake2_256 versus allowing other hashers?

arkpar replied:
Content needs to be ipfs-multihash addressable and hashing scheme would need to be exposed to the IPFS handler somehow. To keep things simple I've hardcoded it in this initial version. This may be changed later if need.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
+				<TransactionRoots<T>>::remove_prefix(obsolete);
+				<TransactionCount<T>>::remove(obsolete);
```
is there an alternative to doing this in the `on_finalize`?

I see that you do account for this properly in the `on_initialize` weight, but even still, we really want to avoid logic in `on_finalize` if we can avoid it.

shawntabrizi replied:
For example, if we are just cleaning things up, why not clean it up at the beginning of the block?

arkpar replied:
> we really want to avoid logic in on_finalize if we can avoid it.

Why? If I don't clean up temporary values in `on_finalize` they'll end up being committed to the database. Something I'd like to avoid.

arkpar replied:
Ah, I can cleanup obsolete data in `on_initialize` indeed.

I'm still curious why `on_initalize` is prefered to  `on_finalize`? Until very recently, `on_initalize` was needed to be called before executing not just blocks, but some runtime APIs. So i'd expected we'd want to put as little as possible there.

shawntabrizi replied:
@arkpar on initialize is preferred because it all happens at the beginning of the block before transactions are included.

This means that for things like PoV size, we will know at the beginning of a block, before we add txs, how much is used and that most probably it will fit because there is nothing else used in the space.

The runtime can handle situations where transactions are not included, but the runtime CANNOT handle not including the `on_finalize` or `on_initialize`. So while the block is empty and we can most safely do all operations, we want to stick it in the front.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			<Counter<T>>::kill();
+			// Drop obsolete roots
+			let period = <StoragePeriod<T>>::get();
+			let obsolete = n.saturating_sub(period);
+			if obsolete > Zero::zero() {
```
should use a checked sub with a match statement.

This will clean up things immediately for the whole first period of a chain, where n is less than period.

arkpar replied:
Not sure what you mean here. `saturating_sub` returns 0 when underflows, so strict check on the following line won't trigger the cleanup. We also don't want to cleanup block 0.

shawntabrizi replied:
ah you are right since obsolete must be greater than zero, not zero itself.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		#[pallet::weight(T::WeightInfo::store(data.len() as u32))]
+		pub(super) fn store(
+			origin: OriginFor<T>,
+			data: Vec<u8>,
```
can use `BoundedVec` which will be type safe for the length

arkpar replied:
Opted to make it runtime configurable, rather then compile time.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			let root = sp_io::trie::blake2_256_ordered_root(chunks);
+
+			let content_hash = sp_io::hashing::blake2_256(&data);
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap();
```
no unwrap in runtime. if you can prove this will never fail, place an `expect` statement.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		pub(super) fn store(
+			origin: OriginFor<T>,
+			data: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
```
```suggestion
		) -> DispatchResult {
```

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+				content_hash: content_hash.into(),
+			});
+			Self::deposit_event(Event::Stored(counter));
+			Ok(().into())
```
```suggestion
			Ok(())
```

And anywhere else you can update this

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			block: T::BlockNumber,
+			index: u32,
+		) -> DispatchResultWithPostInfo {
+			let info = <TransactionRoots<T>>::get(block, index).ok_or(Error::<T>::RenewedNotFound)?;
```
would be nice to move the origin check the top of this function, and then potentially we can fail early without needing to do a storage read.

shawntabrizi replied:
I would say in general, it is a good idea just to move the origin check into each of the extrinsics and then pass sender to `apply_fee`

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			let byte_fee = ByteFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
+			let fee = byte_fee.saturating_mul(size.into());
+			ensure!(T::Currency::can_slash(&sender, fee), Error::<T>::InsufficientFunds);
+			T::Currency::slash(&sender, fee);
```
Should add a imbalance handler, which could allow you to send the fees to treasury, or burn them, or whatever other logic you can think of.

https://github.com/paritytech/substrate/blob/master/frame/treasury/src/lib.rs#L127

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		Blake2_128Concat,
+		T::BlockNumber,
+		Blake2_128Concat,
+		u32,
+		TransactionInfo,
+		OptionQuery,
```
why not a vec here?

arkpar replied:
68 bytes * number of transactions in the block. So a few kilobytes.  I'll change it to Vec as it will probably be faster indeed.

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+			origin: OriginFor<T>,
+			proof: Option<TransactionStorageProof>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
```
What stops other people from calling this function?

arkpar replied:
You mean users making a transaction with this call?
It's an inherent call. Doesn't `ensure_none` effectively check that? I've used the same setup as e.g. `timestamp::set`

shawntabrizi replied:
I think the difference is that `timestamp::set` has a check that timestamp was already set as an early exit, and part of the inherent pre-verification.

I think you can submit an unsigned extrinsic and it will pass the `ensure_none` check, but i dont exactly remember. Would be good to check again, because this function seems super heavy, and nothing is preventing it from running twice (or more) in a block afaik.

shawntabrizi commented on frame/transaction-storage/src/benchmarking.rs:
```diff
@@ -0,0 +1,170 @@
+
+use crate::Pallet as TransactionStorage;
+
+const MAX_EXTRINSICS: u32 = 100;
```
Should be configurable in the pallet. Also I dont see where this is enforced.

arkpar replied:
This is only used in benchmarks

shawntabrizi replied:
But it assumes some data limit about the storage?

If you can only handle 100 extrinsics in benchmarking, i would expect to see some limit to the number of times a user can call the storage pallet in a single block... right?

shawntabrizi commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,364 @@
+		/// # <weight>
+		/// - Linear w.r.t the number of indexed transactions in the proved block for random probing.
+		/// There's a DB read for each transaction.
+		/// Here we assume a maximum of 100 probed transactions.
```
Doesn't seem like  a safe assumption. Should make this limit configurable and enforce it with your counter.

shawntabrizi commented:
Happy to look again when comments above are addressed

arkpar commented:
@shawntabrizi I think I've addressed all the comments. Here's a summary of major changes:

* `TransactionInfo`s are now stored in a vec, instead of a submap. 
* Added runtime configurable limits for max transaction size and max transactions in a block.
* Added checks for proof inherent presence/duplicates
* Moved storage cleanup from `on_finalize` to `on_initialize`
* Added an imbalance handler
* Added fixed fee in addition to per-byte fee.

cheme commented on frame/transaction-storage/src/lib.rs:
```diff
@@ -0,0 +1,436 @@
+	/// Size of indexed data in bytes.
+	size: u32,
+	/// Total number of chunks added in the block with this transaction. This
+	/// is used find transactino info by block chunk index using binary search.
```
```suggestion
	/// is used find transaction info by block chunk index using binary search.
```

arkpar commented:
bot merge

apopiak commented on frame/transaction-storage/README.md:
```diff
@@ -0,0 +1,8 @@
+# Transaction Storage Pallet
+
+Indexes transactions and manages storage proofs.
+# Transaction Storage Pallet
+
+Indexes transactions and manages storage proofs.
+
```
I know I'm late to the party, but it would be very cool to get a bit more of a high level description of the pallet.
